Launcher add poll

cevian opened the pull request.

cevian committed:
Move time_bucket epoch to a Monday

Since Monday is the ISO start of the week, it makes sense to move
the time_bucket epoch to start on a Monday. Before the epoch was the
same as the Postgres epoch (2000-01-01, a Saturday).

cevian committed:
Add a version of time_bucket that takes an origin

This allows people to explicitly specify the origin point.

cevian committed:
Move to using macro for time_bucket_ts

Macro is used for 2 reasons:
1) It's more correct in that it doesn't mix Timestamp and TimestampTz
types. There is no implicit conversion of the two beneath the hood.
2) It is slightly faster as it avoid an extra function call. This
is a very performance sensitive function for OLAP queries.

cevian committed:
Make all time_bucket function STRICT

All time bucket function should return NULL on any NULL parameters.

Commit:
Remove invalid assert in process_altertable_set_tablespace_end

Commit:
Highlight that PG_CONFIG can be used to specify the path to the PostgreSQL installation.

Commit:
Only install the mock functions for debug builds

cevian committed:
Add state machine and polling to launcher

This PR changes the launcher to use a state machine to keep track
of the state of each database scheduler. Further, it add polling
to go through the list of databases and check their states. This
solves several issues

1) A CREATE DATABASE call using a template that already has TimescaleDB
installed previously did not start a scheduler until the next database
restart. A test for this case has been added.
2) A lack of available slots or background workers when a new database was added
meant that the scheduler would not be stared until the next database
restart. Now this will be retried on every polling event.

This PR also simplifies logic since database entries are never removed
from the hash table and thus never added more than once. State
transitions are now easier to read and reason about.

Documentations for the state transitions has been added.

Commit:
Handle ALTER SCHEMA RENAME properly

Previously, we did not intercept the ALTER SCHEMA [name] RENAME command, which meant all Timescale catalog tables that store a schema name were not getting updated properly. This caused problems when users tried to drop a hypertable in a renamed schema, but also would have also caused problems with other commands (which we now add tests for in this PR).

svenklemm committed:
Change return value of create_hypertable to TABLE

Change create_hypertable to return a record consisting of
(hypertable_id, schema_name, table_name). This improves user feedback
about success of the operation but also gives the function an API
returning useful information for non-human consumption.

amytai commented on src/loader/bgw_launcher.c:
```diff
@@ -492,32 +522,42 @@ register_scheduler_handle(VirtualTransactionId vxid, DbHashEntry *db_he)
 
-	if (get_background_worker_pid(db_he->db_scheduler_handle, &worker_pid) == BGWH_STOPPED)
-		return register_scheduler_handle(vxid, db_he);
+	scheduler_state_advance(entry);
```
I think this and the `scheduler_state_advance` in `restart_action` can be eliminated. That way, messages can just get the scheduler into a state that will be ready to be advanced inside the launcher's main loop.

cevian replied:
I don't think so because the return code depends on whether the bgw was able to be relaunched

amytai commented on src/loader/bgw_launcher.c:
```diff
@@ -312,78 +345,138 @@ populate_database_htab(HTAB *db_htab)
+				scheduler_state_trans_allocated_to_start(entry);
+			return;
+		case ALLOCATED:
+			scheduler_state_trans_allocated_to_start(entry);
```
missing return

cevian replied:
good catch

amytai commented on src/loader/bgw_launcher.c:
```diff
@@ -76,8 +95,13 @@ typedef struct DbHashEntry
 													 * properly */
+	SchedulerState state;
+	VirtualTransactionId vxid;
+	int			state_advance_failures;
```
what is this for?

cevian replied:
This is for the error handling code to prevent repeating error messages on every poll

amytai commented on src/loader/bgw_launcher.c:
```diff
@@ -531,24 +571,29 @@ message_stop_action(HTAB *db_htab, BgwMessage *message)
 
-	db_he = hash_search(db_htab, &message->db_oid, HASH_FIND, &found);
-	if (!found)
+	entry = db_hash_entry_create_if_not_exists(db_htab, message->db_oid);
```
I think this changes the current semantics of `restart`, since now if the entry doesn't exist, we create one for it and immediately move it to ENABLED or ALLOCATED.

cevian replied:
true. will fix

amytai commented on src/loader/bgw_launcher.c:
```diff
@@ -492,32 +522,42 @@ register_scheduler_handle(VirtualTransactionId vxid, DbHashEntry *db_he)
-		wait_for_background_worker_shutdown(db_he->db_scheduler_handle);
+		case ENABALED:
+			scheduler_state_trans_enabled_to_disabled(entry);
+			return ACK_SUCCESS;
```
just have one `return ACK_SUCCESS` at the end

cevian replied:
Fixed

amytai reviewed (commented):
I'm generally OK with how the transitions ended up looking. Left some comments, but will do a full review when it's finalized / submitted as PR against master.

RobAtticus committed:
Update code comments in adaptive chunking; mark as beta

We want it to be clear that the adaptive chunking feature is
currently in beta, so a notice at the top of the C file should
suffice even if the user does not see the docs. Also while I was
there I commented some of the functions without comments.

did-g committed:
Search for pg_isolation_regress in pg_config bin directory

svenklemm committed:
Change return value of add_dimension to TABLE

Change the return value of add_dimension to return a record consisting
of dimension_id, schema_name, table_name, column_name. This improves
user feedback about success of the operation but also gives the function
an API returning useful information for non-human consumption.
